File: kattis/ceiling.cpp
#include <bits/stdc++.h>

using namespace std;

int n, k;

struct Tree {
  int value = -1;
  Tree() {}
  Tree(int v) : value(v) {}
  unique_ptr<Tree> left, right;

  void insert(int node) {
    if (value <= node) {
      if (right)
        right->insert(node);
      else
        right = make_unique<Tree>(node);
    } else {
      if (left)
        left->insert(node);
      else
        left = make_unique<Tree>(node);
    }
  }

  uint64_t hash(int factor) {
    uint64_t total = 0;

    if (left)
      total += 0x7c2123c * left->hash(factor + 1) * factor;
    else
      total += 0xc343455 * factor;

    if (right)
      total += 0x919113 * right->hash(factor + 2) * factor;
    else
      total += 0x2af4343 * factor;

    return total;
  }
};

void solve() {
  int node;
  unordered_set<uint64_t> trees;
  for (int i = 0; i < n; i++) {
    Tree t;
    cin >> t.value;
    for (int j = 1; j < k; j++) {
      cin >> node;
      t.insert(node);
    }

    trees.emplace(t.hash(1));
  }

  cout << trees.size() << endl;
}

int main() {
  while (scanf("%d %d", &n, &k) == 2) solve();
  return 0;
}

File: kattis/ceiling.py
class TreeNode:
    value = None
    left = None
    right = None

    def __init__(self, v):
        self.value = v

    def hash(self, factor):
        total = 0

        if self.left:
            total += 0x4234 * factor * self.left.hash(factor * 3)
        else:
            total += 0x434347 * factor

        if self.right:
            total += 0x4114 * factor * self.right.hash(factor * 2)
        else:
            total += 0x1911913 * factor

        return total

    def insert(self, v):
        if v <= self.value:
            if self.left:
                self.left.insert(v)
            else:
                self.left = TreeNode(v)
        else:
            if self.right:
                self.right.insert(v)
            else:
                self.right = TreeNode(v)


n, k = map(int, input().split())

tree_shapes = set()

for _ in range(n):
    values = list(map(int, input().split()))

    root = TreeNode(values[0])

    for i in range(1, k):
        root.insert(values[i])

    tree_shapes.add(root.hash(1))

print(len(tree_shapes))
